### PR TITLE
React dev server plugins additional updates

### PR DESCRIPTION
**Pull request by cowboy**
This is an attempt to address the outstanding issues in https://github.com/cypress-io/cypress/pull/17738#issuecomment-902049761, a PR to address https://cypress-io.atlassian.net/browse/UNIFY-278

### General changes
- [x] Rename the `DevServerOptions` type to `DevServerConfig`
- [x] Rename all dev server functions to `devServer`
- [x] Rename the first `devServer` argument (new signature) to `cypressDevServerConfig`
- [x] Rename the second `devServer` argument (new signature) to `devServerConfig`
- [x] Rename all dev server `devServerConfig` types to follow the pattern `Cypress___DevServerConfig` eg. `CypressBabelDevServerConfig`
- [x] Update each react dev server module to support both the current (legacy) signature as the default export _and_ the new signature as a `devServer` named export
- [x] Update types in each react dev server module `index.d.ts` to support the default + named exports
- [x] Export the type for all dev server `devServerConfig`
- [x] Export a `defineDevServerConfig` helper function in each dev server

### Module-specific changes
- [x] Update the named export CRACO dev server signature second argument to be an object like `{ cracoConfig }`

#### Babel

Legacy (plugins/index.js)
```js
const setupBabelDevServer = require('@cypress/react/plugins/babel')
module.exports = (on, config) => {
  setupBabelDevServer(on, config)
  return config
}
```

New (cypress.config.js)
```js
const { devServer } = require('@cypress/react/plugins/babel')
module.exports = {
  component: {
    devServer,
    devServerConfig: {/* options */}
  }
}
```

#### CRACO

Legacy (plugins/index.js)
```js
const setupCracoDevServer = require('@cypress/react/plugins/craco')
const cracoConfig = require('../../craco.config.js')
module.exports = (on, config) => {
  setupCracoDevServer(on, config, cracoConfig)
  return config
}
```

New (cypress.config.js)
```js
const { devServer } = require('@cypress/react/plugins/craco')
const cracoConfig = require('./craco.config.js')
module.exports = {
  component: {
    devServer,
    devServerConfig: { cracoConfig }
  }
}
```

#### Load-webpack

Legacy (plugins/index.js)
```js
const setupWebpackDevServer = require('@cypress/react/plugins/load-webpack')
module.exports = (on, config) => {
  setupWebpackDevServer(on, config, {
    webpackFilename: 'webpack.config.js',
  })
  return config
}
```

New (cypress.config.js)
```js
const { devServer } = require('@cypress/react/plugins/load-webpack')
module.exports = {
  component: {
    devServer,
    devServerConfig: { webpackFilename: 'webpack.config.js' }
  }
}
```

#### Next.js

Legacy (plugins/index.js)
```js
const setupNextDevServer = require('@cypress/react/plugins/next')
module.exports = (on, config) => {
  setupNextDevServer(on, config)
  return config
}
```

New (cypress.config.js)
```js
const { devServer } = require('@cypress/react/plugins/next')
module.exports = {
  component: {
    devServer
  }
}
```

#### react-scripts

Legacy (plugins/index.js)
```js
const setupCRADevServer = require('@cypress/react/plugins/react-scripts')
module.exports = (on, config) => {
  setupCRADevServer(on, config)
  return config
}
```

New (cypress.config.js)
```js
const { devServer } = require('@cypress/react/plugins/react-scripts')
module.exports = {
  component: {
    devServer,
    devServerConfig: {/* options */}
  }
}
```

## Validating types

I wasn't sure if there was a good way to do this automatically, so I did it manually. I wanted to validate that the `index.d.ts` files were correct, so I created the following `npm/react/plugins/test-all-plugins.ts` file and visually verified in VS Code that lines were shown as valid or invalid. I didn't commit the file, but I'm referencing it here in case it helps review the PR.

```ts
import legacyBabelDevServer, { devServer as babelDevServer } from './babel'
import legacyCracoDevServer, { devServer as cracoDevServer } from './craco'
import legacyWebpackDevServer, { devServer as webpackDevServer } from './load-webpack'
import legacyNextDevServer, { devServer as nextDevServer } from './next'
import legacyCRADevServer, { devServer as craDevServer } from './react-scripts'

import type { CypressBabelDevServerConfig } from "./babel"
import type { CypressCracoDevServerConfig } from "./craco"
import type { CypressWebpackDevServerConfig } from "./load-webpack"
import type { CypressCRADevServerConfig } from "./react-scripts"

const cypressBabelDevServerConfigs: CypressBabelDevServerConfig[] = [
  // valid
  {},
  { setWebpackConfig(config) { return config } },
  // invalid
  { foo: 123 },
  { setWebpackConfig(config) { return 123 } }
]

const cypressCracoDevServerConfigs: CypressCracoDevServerConfig[] = [
  // valid
  { cracoConfig: { 'i am': 'a craco config' } },
  // invalid
  {},
  { 'i am': 'a craco config' }
]

const cypressWebpackDevServerConfigs: CypressWebpackDevServerConfig[] = [
  // valid
  {},
  { webpackFilename: 'foo' },
  // invalid
  { foo: 123 }
]

const cypressCRADevServerConfigs: CypressCRADevServerConfig[] = [
  // valid
  {},
  { webpackConfigPath: 'foo' },
  // invalid
  { foo: 123 },
  { webpackConfigPath: 123 }
]

const pluginIndex = (on: Cypress.PluginEvents, config: Cypress.PluginConfigOptions) => {
  // valid
  legacyBabelDevServer(on, config)
  legacyBabelDevServer(on, config, {})
  legacyBabelDevServer(on, config, { setWebpackConfig(config) { return config } })
  // invalid
  legacyBabelDevServer()
  legacyBabelDevServer(on, config, { foo: 123 })
  legacyBabelDevServer(on, config, { setWebpackConfig(config) { return 123 } })

  // valid
  legacyCracoDevServer(on, config, { 'i am': 'a craco config' })
  // invalid
  legacyCracoDevServer()
  legacyCracoDevServer(on, config)

  // valid
  legacyWebpackDevServer(on, config)
  legacyWebpackDevServer(on, config, {})
  legacyWebpackDevServer(on, config, { webpackFilename: 'foo' })
  // invalid
  legacyWebpackDevServer()
  legacyWebpackDevServer(on, config, { foo: 123 })
  legacyWebpackDevServer(on, config, { webpackFilename: 123 })

  // valid
  legacyNextDevServer(on, config)
  legacyNextDevServer(on, config, {})
  // invalid
  legacyNextDevServer()

  // valid
  legacyCRADevServer(on, config)
  legacyCRADevServer(on, config, {})
  legacyCRADevServer(on, config, { webpackConfigPath: 'foo' })
  // invalid
  legacyCRADevServer()
  legacyCRADevServer(on, config, { foo: 123 })
  legacyCRADevServer(on, config, { webpackConfigPath: 123 })
}

const startDevServer = (cypressDevServerConfig: Cypress.DevServerConfig) => {
  // valid
  babelDevServer(cypressDevServerConfig)
  babelDevServer(cypressDevServerConfig, {})
  babelDevServer(cypressDevServerConfig, { setWebpackConfig(config) { return config } })
  // invalid
  babelDevServer()
  babelDevServer(cypressDevServerConfig, { foo: 123 })
  babelDevServer(cypressDevServerConfig, { setWebpackConfig(config) { return 123 } })

  // valid
  cracoDevServer(cypressDevServerConfig, { cracoConfig: { 'i am': 'a craco config' }})
  // invalid
  cracoDevServer()
  cracoDevServer(cypressDevServerConfig)
  cracoDevServer(cypressDevServerConfig, {})
  cracoDevServer(cypressDevServerConfig, { 'i am': 'a craco config' })

  // valid
  webpackDevServer(cypressDevServerConfig)
  webpackDevServer(cypressDevServerConfig, {})
  webpackDevServer(cypressDevServerConfig, { webpackFilename: 'foo' })
  // invalid
  webpackDevServer()
  webpackDevServer(cypressDevServerConfig, { foo: 123 })
  webpackDevServer(cypressDevServerConfig, { webpackFilename: 123 })

  // valid
  nextDevServer(cypressDevServerConfig)
  nextDevServer(cypressDevServerConfig, {})
  // invalid
  nextDevServer()

  // valid
  craDevServer(cypressDevServerConfig)
  craDevServer(cypressDevServerConfig, {})
  craDevServer(cypressDevServerConfig, { webpackConfigPath: 'foo' })
  // invalid
  craDevServer()
  craDevServer(cypressDevServerConfig, { foo: 123 })
  craDevServer(cypressDevServerConfig, { webpackConfigPath: 123 })
}
```

### Future work

(This has been documented in https://cypress-io.atlassian.net/browse/UNIFY-295)

When we deprecate the legacy APIs/signatures:

- Add a deprecation warning into the `getLegacyDevServer` function that warns when it is called (not when the returned function is called) **OR**
- Update the test runner to add a deprecation warning into the code that requires the plugins/index.js file

When we remove the legacy APIs/signatures:

- Remove the `getLegacyDevServer` function
- Refactor each dev server from this (roughly):

```js
// Legacy signature
module.exports = getLegacyDevServer(devServer, (config) => {
  // code
})

// New signature
module.exports.devServer = devServer
```

To this:

```js
module.exports = { devServer }
```

(Also note that the CRACO `devServer` function will need to be updated to accept the proper `devServerConfig` object)

- Refactor `index.d.ts` types from this (roughly):

```ts

declare namespace legacyDevServer {
  interface CypressFooDevServerConfig {
    // properties
  }

  export function devServer(/* arguments */): Cypress.ResolvedDevServerConfig
}

declare function legacyDevServer(/* arguments */): void

export = legacyDevServer;
```

To this:

```ts
export interface CypressFooDevServerConfig {
  // properties
}

export function devServer(/* arguments */): Cypress.ResolvedDevServerConfig
```

### CODE REVIEW

**Review by lmiller1990 — COMMENTED** [review]
All the tests are still passing, so the changes are not breaking anything. Have we tried the new react plugins with the new cypress.config.js yet (not sure if we actually implemented `startDevServer` yet?)

Re: types, since the files that contain the exports (eg `npm/react/plugins/babel/index.js` etc) are pretty simple, rather than trying to manually sync/validate types, should we just change them to TS and avoid that all together? Some files in `npm/react` are already using TS, so we could just modify how we build the files so that the `js` file is alongside the `ts` file, eg:

- `npm/react/plugins/babel/index.ts` => `yarn tsc` => creates `npm/react/plugins/babel/index.js`

Thoughts?

There's also a library, `tsd`, that can check types: https://github.com/SamVerschueren/tsd. I've used it in Vue Test Utils successfully, since we have some fairly complex generic types: https://github.com/vuejs/vue-test-utils-next/tree/master/test-dts

File: npm/react/plugins/craco/index.d.ts [hunk 1]
```diff
@@ -1,16 +1,26 @@
+interface CracoDevServerOptions {
```
Comment (cowboy) [inline npm/react/plugins/craco/index.d.ts:1]: Rename from `CracoDevServerOptions` -> `CypressCracoDevServerConfig`

File: npm/react/plugins/babel/index.d.ts [hunk 2]
```diff
@@ -10,6 +10,11 @@ declare namespace legacyDevServer {
     setWebpackConfig?(config:Configuration): Configuration
   }
 
+  /**
+   * Type helper to make writing `CypressBabelDevServerConfig` easier
+   */
+  export function defineDevServerConfig(babelDevServerConfig: CypressBabelDevServerConfig): CypressBabelDevServerConfig
```
Comment (cowboy) [inline npm/react/plugins/babel/index.d.ts]: For consistency, the first argument to `defineDevServerConfig` should be called `devServerConfig` in all modules.

File: npm/react/plugins/babel/index.js [hunk 3]
```diff
@@ -18,3 +18,7 @@ module.exports = getLegacyDevServer(devServer, (config) => {
 
 // New signature
 module.exports.devServer = devServer
+
+module.exports.defineDevServerConfig = function (config) {
```
Comment (cowboy) [inline npm/react/plugins/babel/index.js:22]: For consistency, the first argument to defineDevServerConfig should be called devServerConfig in all modules.

**Review by cowboy — COMMENTED** [review]
@elevatebart a few comments, but otherwise looks good!

**Pull request merged**